app/src/main/java/com/github/onlynotesswent/utils/TextExtractor.kt

```diff
@@ -0,0 +1,102 @@
+package com.github.onlynotesswent.utils
+
+import android.graphics.Bitmap
+import android.graphics.pdf.PdfRenderer
+import android.os.ParcelFileDescriptor
+import android.util.Log
+import android.widget.Toast
+import androidx.activity.ComponentActivity
+import com.google.mlkit.vision.common.InputImage
+import com.google.mlkit.vision.text.TextRecognition
+import com.google.mlkit.vision.text.TextRecognizer
+import com.google.mlkit.vision.text.latin.TextRecognizerOptions
+import java.io.File
+
+/**
+ * TextExtractor class that detects latin-based text from a PDF file. The class uses the Google ML
+ * Kit Text Recognition API to extract text from the PDF file.
+ *
+ * @param activity the ComponentActivity that will use the TextExtractor
+ */
+class TextExtractor(private val activity: ComponentActivity) {
+  private lateinit var textRecognizer: TextRecognizer
+
+  fun init() {
+    textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
+  }
+
+  /**
+   * Processes the given PDF file to extract text from it.
+   *
+   * @param pdfFile the PDF file to extract text from.
+   * @param onSuccess the lambda function to call with the extracted text.
+   */
+  fun processPdfFile(pdfFile: File, onSuccess: (String) -> Unit) {
+    if (!::textRecognizer.isInitialized) {
+      // Case should not happen if class is correctly initialized
+      Log.e(TAG, "TextRecognizer is not initialized. Call init() before processing PDF file.")
+      return
+    }
+
+    try {
+      val bitmaps = convertPdfToBitmap(pdfFile)
+      extractTextFromBitmaps(bitmaps, onSuccess)
+    } catch (e: Exception) {
+      Log.e(TAG, "Error while converting PDF to bitmap", e)
+      Toast.makeText(activity, "Error: text recognition failed", Toast.LENGTH_SHORT).show()
+    }
+  }
+
+  /**
+   * Converts the given PDF file to a list of bitmaps.
+   *
+   * @param pdfFile the PDF file to convert to bitmaps
+   * @return a list of bitmaps extracted from the PDF file
+   * @throws Exception if the PDF file cannot be converted to bitmaps
+   */
+  private fun convertPdfToBitmap(pdfFile: File): List<Bitmap> {
+    val bitmaps = mutableListOf<Bitmap>()
+    val fileDescriptor = ParcelFileDescriptor.open(pdfFile, ParcelFileDescriptor.MODE_READ_ONLY)
+    val pdfRenderer = PdfRenderer(fileDescriptor)
+
+    for (pageIndex in 0 until pdfRenderer.pageCount) {
+      val page = pdfRenderer.openPage(pageIndex)
+      val bitmap = Bitmap.createBitmap(page.width, page.height, Bitmap.Config.ARGB_8888)
+      page.render(bitmap, null, null, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
+      bitmaps.add(bitmap)

```

Comment:
I think that you should use bitmap.recycle() here to free some memory. I am afraid we could have some potential memory issues if this not done. I think that the bitmap won't be properly garbage collected here as it is added to the bitmaps list.

Reply:
Yes, you are right. I will also fix the issue where some values were not closed properly in case an exception is thrown. Thank you!